parser.go

```diff
@@ -422,19 +429,85 @@ func (r *Rule) network(key item, l *lexer) error {
 	}
 	switch key.typ {
 	case itemSourceAddress:
-		r.Source.Nets = append(r.Source.Nets, items...)
+		if validNetworks(items) {
+			r.Source.Nets = append(r.Source.Nets, items...)
+		} else {
+			return fmt.Errorf("some or all source ips are invalid: %v", items)
+		}
 	case itemSourcePort:
-		r.Source.Ports = append(r.Source.Ports, items...)
+		if portsValid(items) {
+			r.Source.Ports = append(r.Source.Ports, items...)
+		} else {
+			return fmt.Errorf("some or all source ports are invalid: %v", items)
+		}
 	case itemDestinationAddress:
-		r.Destination.Nets = append(r.Destination.Nets, items...)
+		if validNetworks(items) {
+			r.Destination.Nets = append(r.Destination.Nets, items...)
+		} else {
+			return fmt.Errorf("some or all destination ips are invalid: %v", items)
+		}
 	case itemDestinationPort:
-		r.Destination.Ports = append(r.Destination.Ports, items...)
+		if portsValid(items) {
+			r.Destination.Ports = append(r.Destination.Ports, items...)
+		} else {
+			return fmt.Errorf("some or all destination ports are invalid: %v", items)
+		}
 	default:
 		panic("item is not a network component")
 	}
 	return nil
 }
 
+// Validate that every item is between 1 and 65535.
+func portsValid(p []string) bool {
+	for _, _port := range p {
+
+		if len(_port) > 0 && string(_port[0:1]) == "!" {
+			_port = _port[1:]
+		}
+		_ports := strings.Split(_port, ":")
+		for _, port := range _ports {
+			if port == "any" || strings.HasPrefix(port, "$") {
+				continue
+			}
+			x, err := strconv.Atoi(port)
+			if err != nil {
+				return false
+			}
+			if x > 65535 || x < 1 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// Validate item is either a valid ip or ip range.
+func validNetwork(ip string) bool {
```

Comment:
you might be able to avoid some additional calls by checkin ParseIP and ParseCIDR independently something like this:

https://play.golang.org/p/TeAkkPVlz-j

If rules in general show that CIDR networks are more common than IPs then swapping the order to return first on a valid CIDR value would be the faster option.

Reply:
done